core/src/main/scala/org/allenai/common/ParIterator.scala

```diff
@@ -0,0 +1,58 @@
+package org.allenai.common
+
+import java.util.concurrent.Semaphore
+
+import scala.concurrent._
+import scala.concurrent.duration.Duration
+
+object ParIterator {
+  val queueLimit = 1024
+
+  implicit class ParIteratorEnrichment[T](val input: TraversableOnce[T]) extends AnyVal {
+
+    def parForeach(f: T => Unit)(implicit ec: ExecutionContext): Unit = {
+      // If there are a billion items in the iterator, we don't want to create a billion futures,
+      // so we limit the number of futures we create with this semaphore.
+      val sema = new Semaphore(queueLimit)
+      input.foreach { item =>
+        // Try to pass it off to a future. If no futures are available, do the work
+        // in this thread.
+        val success = sema.tryAcquire()
+        if (success)
+          Future {
+            f(item)
+            sema.release()
```

Comment:
I did this just now, but this points to a much bigger problem. If an exception happens in `f`, I have to import that exception to the calling thread and throw it there. In other words, I have to call `get` on each of those futures, which means I'll have to change the whole way `parForeach` works.